EIPS/eip-3540.md

```diff
@@ -90,91 +87,99 @@ The list of section headers is terminated with the *section headers terminator b
 
+**NOTE**: A special value of `outputs` being `0x80` is designated to denote non-returning functions as defined in a separate EIP.
+
 #### EOF version 1 validation rules
 
-1. In addition to general validation rules above, EOF version 1 bytecode conforms to the rules specified below:
-   - Exactly one type section header MUST be present immediately following the EOF version. Each code section MUST have a specified type signature in the type body.
-   - Exactly one code section header MUST be present immediately following the type section. A maximum of 1024 individual code sections are allowed.
-   - Exactly one data section header MUST be present immediately following the code section.
-2. Any version other than `0x01` is invalid.
+The following validity constraints are placed on the container format:
 
-(*Remark:* Contract creation code SHOULD set the section size of the data section so that the constructor arguments fit it.)
+- `version` must be `0x01`
+- `types_size` is divisible by `4`
+- the number of code sections must be equal to `types_size / 4`
+- the number of code sections must be greater than `0` and not exceed `1024`
+- `code_size` may not be `0`
+- the number of container sections must not exceed `256`. The number of container sections may not be `0`, if declared in the header
+- `container_size` may not be `0`
+- data section is mandatory, but `data_size` may be `0`
+- data body length may be shorter than `data_size` for a not yet deployed container
 
 ### Changes to execution semantics
 
-For clarity, the *container* refers to the complete account code, while *code* refers to the contents of the code section only.
-
-1. Execution starts at the first byte of the first code section, and PC is set to 0.
-2. Execution stops if `PC` goes outside the code section bounds.
-3. `PC` returns the current position within the *code*.
-4. `CODECOPY`/`CODESIZE`/`EXTCODECOPY`/`EXTCODESIZE`/`EXTCODEHASH` keeps operating on the entire *container*.
-5. The input to `CREATE`/`CREATE2` is still the entire *container*.
-6. The size limit for deployed code as specified in [EIP-170](./eip-170.md) and for initcode as specified in [EIP-3860](./eip-3860.md) is applied to the entire *container* size, not to the *code* size. This also means if initcode validation fails, it is still charged the EIP-3860 `initcode_cost`.
-7. When an EOF1 contract performs a `DELEGATECALL` the target must be EOF1. If it is not EOF1, the `DELEGATECALL` execution finishes as a failed call by pushing `0` to the stack. Only initial gas cost of `DELEGATECALL` is consumed (similarly to the call depth check) and the target address still becomes warm.
+For an EOF contract:
 
-(*Remark:* Due to [EIP-4750](./eip-4750.md), `JUMP` and `JUMPI` are disabled and therefore are not discussed in relation to EOF.)
+- Execution starts at the first byte of code section 0, and PC is set to 0.
+- `PC` is scoped to the executing code section
```

Comment:
Probably this also can be removed.

Reply:
right, but isn't it a remark on how the context variable PC works, rather than opcode? I never was really sure. If you think opcode - let's remove. Actually, I think EIP-4750 elaborates enough on how PC works as a context variable, so I guess we can remove anyway.

Reply:
I thought so too at first, but it's next to the list of what happens to other opcodes, so very much looks like about opcode too.

Reply:
088f5e25 a final check on this pls. I think this leaves PC out completely as irrelevant, making 3540 define how EOF would have worked without CALLF/RETF (only 0th code section executes), which I think is coherent.